File: core/stats.py
class Stats:
    """Combat caps and base attributes.

    Keep transient values (HP/MP) in Health/Mana, not here.
    """

    def __init__(self, max_hp, atk, defense, speed, mp_max=10):
        self.max_hp = int(max_hp)
        self.atk = int(atk)
        self.defense = int(defense)
        self.speed = int(speed)
        self.mp_max = int(mp_max)

    def __str__(self):
        return (
            f"Stats(hp={self.max_hp}, atk={self.atk}, "
            f"def={self.defense}, spd={self.speed}, mp_max={self.mp_max})"
        )


class Health:
    def __init__(self, current, max):
        self.max = int(max)
        self.current = int(current)
        self.clamp()

    def clamp(self):
        if self.current < 0:
            self.current = 0
        elif self.current > self.max:
            self.current = self.max

    def is_dead(self):
        return self.current <= 0

    def ratio(self):
        return 0.0 if self.max <= 0 else self.current / self.max

    def __str__(self):
        return f"Health({self.current}/{self.max})"


class Mana:
    def __init__(self, current=0, max=10):
        self.max = int(max)
        self.current = int(current)
        self.clamp()

    def clamp(self):
        if self.current < 0:
            self.current = 0
        elif self.current > self.max:
            self.current = self.max

    def full(self):
        return self.current >= self.max

    def ratio(self):
        return 0.0 if self.max <= 0 else self.current / self.max

    def __str__(self):
        return f"Mana({self.current}/{self.max})"

File: web/src/domain/stats.ts
export class Stats {
  maxHp: number;
  atk: number;
  defense: number;
  speed: number;
  mpMax: number;

  constructor(params: {
    maxHp: number;
    atk: number;
    defense: number;
    speed: number;
    mpMax?: number;
  }) {
    this.maxHp = toInt(params.maxHp);
    this.atk = toInt(params.atk);
    this.defense = toInt(params.defense);
    this.speed = toInt(params.speed);
    this.mpMax = toInt(params.mpMax ?? 10);
  }

  toString(): string {
    return `Stats(hp=${this.maxHp}, atk=${this.atk}, def=${this.defense}, spd=${this.speed}, mp_max=${this.mpMax})`;
  }
}

export class Health {
  max: number;
  current: number;

  constructor(params: { current: number; max: number }) {
    this.max = toInt(params.max);
    this.current = toInt(params.current);
    this.clamp();
  }

  clamp(): void {
    if (this.current < 0) this.current = 0;
    else if (this.current > this.max) this.current = this.max;
  }

  isDead(): boolean {
    return this.current <= 0;
  }

  ratio(): number {
    return this.max <= 0 ? 0.0 : this.current / this.max;
  }

  toString(): string {
    return `Health(${this.current}/${this.max})`;
  }
}

export class Mana {
  max: number;
  current: number;

  constructor(params: { current?: number; max?: number } = {}) {
    this.max = toInt(params.max ?? 10);
    this.current = toInt(params.current ?? 0);
    this.clamp();
  }

  clamp(): void {
    if (this.current < 0) this.current = 0;
    else if (this.current > this.max) this.current = this.max;
  }

  full(): boolean {
    return this.current >= this.max;
  }

  ratio(): number {
    return this.max <= 0 ? 0.0 : this.current / this.max;
  }

  toString(): string {
    return `Mana(${this.current}/${this.max})`;
  }
}

function toInt(value: number): number {
  return Math.trunc(value);
}
